### PR TITLE
Blog Post: Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models 

### LINKED ISSUE #3164: Blog Post: Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models
**Describe the blog post your would like to write**
An introduction and performance benchmark to the Neural Sparse v2 models.

**What is the title of the blog post?**
Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models

**Who are the authors?**
Zhichao Geng, Cong Guan, Charlie Yang

**What is the proposed posting date?**
In August

### PR DESCRIPTION
**Pull request by zhichao-aws**
### Description
Create blog: Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models 
 
### Issues Resolved
#3164 

### Check List
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

**Comment by zhichao-aws** [comment]
The blog should be released after the update to current documentation on supported sparse models. I'll give update on that PR here

**Comment by zhichao-aws** [comment]
The PR link for documentation: https://github.com/opensearch-project/documentation-website/pull/7987

**Comment by kolchfa-aws** [comment]
@zhichao-aws The [doc PR](https://github.com/opensearch-project/documentation-website/pull/7987) is merged.

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        "name": "amazon/neural-sparse/opensearch-neural-sparse-encoding-doc-v2-distill",
+        "version": "1.0.0",
+        "model_format": "TORCH_SCRIPT"
+    }
+    ```
+
+1. Register and deploy a tokenizer for search:
+
+    ```
```
Comment (natebower) [inline _posts/2024-08-19-neural-sparse-v2-models.md:133]: json?

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        "name": "amazon/neural-sparse/opensearch-neural-sparse-tokenizer-v1",
+        "version": "1.0.1",
+        "model_format": "TORCH_SCRIPT"
+    }
+    ```
+
+1. Get model IDs for the model and tokenizer by calling the Tasks API:
+
+    ```
```
Comment (natebower) [inline _posts/2024-08-19-neural-sparse-v2-models.md:144]: json?

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        "name": "amazon/neural-sparse/opensearch-neural-sparse-encoding-v2-distill",
+        "version": "1.0.0",
+        "model_format": "TORCH_SCRIPT"
+    }
+    ```
+
+1. Get the model ID for the sparse encoding model by calling the Tasks API:
+
+    ```
```
Comment (natebower) [inline _posts/2024-08-19-neural-sparse-v2-models.md:165]: json?

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    GET /_plugins/_ml/tasks/{task_id}
+    ```
+
+## Further reading
+
+For more information about neural sparse search, see our other blog posts:
```
Comment (natebower) [inline _posts/2024-08-19-neural-sparse-v2-models.md:171]:
```suggestion
For more information about neural sparse search, see these previous blog posts:
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+- [Improving document retrieval with sparse semantic encoders]({{site.baseurl}}/blog/improving-document-retrieval-with-sparse-semantic-encoders)
+- [A deep dive into faster semantic sparse retrieval in OpenSearch 2.12]({{site.baseurl}}/blog/A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12)
+- [Introducing the neural sparse two-phase algorithm]({{site.baseurl}}/blog/Introducing-a-neural-sparse-two-phase-algorithm)
+
+---
+
+[^1]: For pretraining, we selected a portion of the sentence transformers [training data](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2#training-data). We removed any data that also appeared in BEIR to maintain a zero-shot evaluation environment.
```
Comment (natebower) [inline _posts/2024-08-19-neural-sparse-v2-models.md:179]:
```suggestion
[^1]: For pretraining, we selected a portion of the sentence transformer [training data](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2#training-data). We removed any data that also appeared in BEIR in order to maintain a zero-shot evaluation environment.
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+featured_blog_post: true 
+featured_image: false # /assets/media/blog-images/__example__image__name.jpg
+---
+
+Neural sparse search is a novel and efficient method for semantic retrieval, [introduced in OpenSearch 2.11](https://opensearch.org/blog/improving-document-retrieval-with-sparse-semantic-encoders/). Sparse encoding models encode text into (token, weight) entries, allowing OpenSearch to build indexes and perform searches using Lucene's inverted index. Neural sparse search is efficient and generalizes well in out-of-domain (OOD) scenarios. We are excited to announce the release of our v2 series neural sparse models:
+
+- **v2-distill model**: This model **reduces model parameters by 50%**, resulting in lower memory requirements and costs. It **increases ingestion throughput by 1.39 on GPU and 1.74x on CPU**. The v2-distill architecture supports both doc-only and bi-encoder modes.
+- **v2-mini model**: This model **reduces model parameters by 75%**, also reducing memory requirements and costs. It **increases ingestion throughput by 1.74x on GPU and 4.18x on CPU**. The v2-mini architecture supports the doc-only mode.
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:25]:
```suggestion
- **v2-mini model**: This model **reduces model parameters by 75%**, also reducing memory requirements and costs. It **increases ingestion throughput by 1.74x on GPUs and 4.18x on CPUs**. The v2-mini architecture supports the doc-only mode.
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+Dense retrievers are usually pretrained using Information Noise-Contrastive Estimation (InfoNCE) loss, which improves consistency and uniformity of dense representations. However, we found that InfoNCE loss does not enhance sparse embeddings in doc-only mode as it does for dense models. Instead, knowledge distillation loss is a more effective optimization technique for sparse encoding models. Finding a suitable teacher model for large-scale pretraining is challenging. Siamese encoders are generally not strong models, while cross-encoders struggle with handling large pretraining datasets. Inspired by the performance boost from the [hybrid search of dense and lexical (sparse) approaches](https://opensearch.org/blog/hybrid-search/), we decided to combine a bi-encoder sparse retriever with Siamese dense models to create a strong teacher model. This model combines the strengths of heterogeneous retrievers and is efficient enough for pretraining. We plan to publish a paper detailing the training procedure.
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:54]:
```suggestion
Dense retrievers are usually pretrained using Information Noise-Contrastive Estimation (InfoNCE) loss, which improves consistency and uniformity of dense representations. However, we found that InfoNCE loss does not enhance sparse embeddings in doc-only mode as it does for dense models. Instead, knowledge distillation loss is a more effective optimization technique for sparse encoding models. Finding a suitable teacher model for large-scale pretraining is challenging. Siamese encoders are generally not strong models, while cross-encoders struggle with handling large pretraining datasets. Inspired by the performance boost demonstrated by the dense and lexical (sparse) [hybrid search](https://opensearch.org/blog/hybrid-search/), we decided to combine a bi-encoder sparse retriever with Siamese dense models to create a strong teacher model. This model combines the strengths of heterogeneous retrievers and is efficient enough for pretraining. We plan to publish a paper detailing the training procedure.
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+In these experiments, we set the `batch_size` of the `sparse_encoding` ingestion processor to `2`. We recorded the mean ingestion throughput and the p99 client-side latency for the Bulk API, using 20 clients for ingestion.
+
+#### Remote deployment using GPU 
+
+The bulk size was set to 24. The experiment results are listed in the following figure. Compared with the v1 model, the **v2-distill** model provided a **1.39x increase in mean throughput** and the **v2-mini** model provided a **1.74x increase in mean throughput**.
+
+<img src="/assets/media/blog-images/2024-08-19-neural-sparse-v2-models/gpu_ingest.png"/>
+
+#### Local deployment using CPU
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:73]:
```suggestion
#### Local deployment on a CPU
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+The bulk size was set to 8. Compared with the v1 model, the **v2-distill** model provided a **1.58x increase in mean throughput** and the **v2-mini** model provided a **4.18x increase in mean throughput**.
+
+<img src="/assets/media/blog-images/2024-08-19-neural-sparse-v2-models/cpu_ingest.png"/>
+
+### Search
+
+In these experiments, we ingested 1 million documents into an index and used 20 clients to perform concurrent searches. We recorded the p99 for both client-side search and model inference. We tested search performance for the **bi-encoder** mode.
+
+#### Remote deployment using GPU
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:83]:
```suggestion
#### Remote deployment on a GPU 
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+In these experiments, we ingested 1 million documents into an index and used 20 clients to perform concurrent searches. We recorded the p99 for both client-side search and model inference. We tested search performance for the **bi-encoder** mode.
+
+#### Remote deployment using GPU 
+
+Compared with the v1 model, the **v2-distill** model **decreased client-side search latency by 11.7% and model inference latency by 23%**.
+
+<img src="/assets/media/blog-images/2024-08-19-neural-sparse-v2-models/gpu_search.png"/>
+
+#### Local deployment using CPU
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:89]:
```suggestion
#### Local deployment on a CPU
```

File: _posts/2024-08-19-neural-sparse-v2-models.md (hunk 1 above)
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:133]:
````suggestion
    ```json
````

File: _posts/2024-08-19-neural-sparse-v2-models.md (hunk 3 above)
Comment (kolchfa-aws) [reply _posts/2024-08-19-neural-sparse-v2-models.md:165]: On the project side, when we don't have the actual request body, just the path and method will not be highlighted.

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+### Knowledge distillation from an ensemble of heterogeneous teacher models
+
+In neural models, performance is closely tied to the number of parameters. When using a smaller architecture, the training algorithm must be enhanced to offset any performance drop. For retrieval tasks, common techniques include **pretraining** and **knowledge distillation**:
+
+- **Pretraining**: This involves training models on large datasets, often constructed using specific rules. Examples of such dataset elements are `(title, body)` pairs in news articles or `(question, answer)` pairs on Q&A websites. Pretraining enhances the model’s search relevance and ability to generalize.
+- **Knowledge distillation**: Some models are powerful but suffer from inefficiencies because of their large size or complex structure. An example of such a model is the cross-encoder reranker. Distillation transfers knowledge from these teacher models to smaller ones, preserving high performance while eliminating these drawbacks.
+
```
[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _posts/2024-08-19-neural-sparse-v2-models.md:53]: Reworded.

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+**GPU deployment**: We used an Amazon SageMaker GPU endpoint to host the neural sparse model, connecting to it using a remote connector. For the complete deployment code, see [this example](https://github.com/zhichao-aws/neural-search/blob/neural_sparse_sagemaker/neural_sparse_sagemaker_example/run.ipynb). The model was hosted on a **g5.xlarge** GPU instance.  
+**CPU deployment**: The model was deployed on **all three nodes** in the cluster.
+
+
+In these experiments, we set the `batch_size` of the `sparse_encoding` ingestion processor to `2`. We recorded the mean ingestion throughput and the p99 client-side latency for the Bulk API, using 20 clients for ingestion.
+
+#### Remote deployment using GPU
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:68]:
```suggestion
#### Remote deployment on a GPU 
```

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+The bulk size was set to 8. Compared with the v1 model, the **v2-distill** model provided a **1.58x increase in mean throughput**, and the **v2-mini** model provided a **4.18x increase in mean throughput**, as shown in the following figure.
+
+<img src="/assets/media/blog-images/2024-08-19-neural-sparse-v2-models/cpu_ingest.png"/>
+
+### Search
+
+In these experiments, we ingested 1 million documents into an index and used 20 clients to perform concurrent searches. We recorded the p99 latency for both client-side search and model inference. We tested search performance for the **bi-encoder** mode.
+
+#### Remote deployment using GPU
```
Comment (kolchfa-aws) [inline _posts/2024-08-19-neural-sparse-v2-models.md:84]:
```suggestion
#### Remote deployment on a GPU 
```

**Comment by kolchfa-aws** [comment]
@pajuric @zhichao-aws Editorial comments are implemented and the blog is ready to publish. Thanks!

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+  - congguan
+  - yych
+  - dylantong
+  - kolchfa
+date: 2024-08-19
+categories:
+    - technical-posts
+has_science_table: true
+meta_keywords: OpenSearch semantic search, neural sparse search, semantic sparse retrieval
+meta_description: Accelerating inference and improving search with v2 neural sparse encoding models
```
Comment (pajuric) [inline _posts/2024-08-19-neural-sparse-v2-models.md:15]: Please update the blog with the following meta: 

meta_keywords: neural sparse models, OpenSearch semantic search,  semantic sparse retrieval, neural search

meta_description: OpenSearch announces the availability of v2 series neural sparse models that enhance the efficiency of semantic sparse retrieval while accelerating inference and improving search.

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
+---
+layout: post
+title:  Improving search efficiency and accuracy with the newest v2 neural sparse models
+authors:
+  - zhichaog
+  - congguan
+  - yych
+  - dylantong
+  - kolchfa
+date: 2024-08-19
```
Comment (pajuric) [inline _posts/2024-08-19-neural-sparse-v2-models.md:10]: Please update the date to 2024-08-21

File: _posts/2024-08-19-neural-sparse-v2-models.md (hunk 14 above)
Comment (kolchfa-aws) [reply _posts/2024-08-19-neural-sparse-v2-models.md:15]: Done

File: _posts/2024-08-19-neural-sparse-v2-models.md (hunk 15 above)
Comment (kolchfa-aws) [reply _posts/2024-08-19-neural-sparse-v2-models.md]: Done

**Comment by pajuric** [comment]
@nateynateynate @krisfreedain - Blog is ready to publish today! Let's ship it.

File: _posts/2024-08-19-neural-sparse-v2-models.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  - dylantong
+  - kolchfa
+date: 2024-08-21
+categories:
+    - technical-posts
+has_science_table: true
+meta_keywords: neural sparse models, OpenSearch semantic search, semantic sparse retrieval, neural search
+meta_description: Accelerating inference and improving search with v2 neural sparse encoding models
+excerpt: OpenSearch announces the availability of v2 series neural sparse models that enhance the efficiency of semantic sparse retrieval while accelerating inference and improving search.
+featured_blog_post: true
```
Comment (krisfreedain) [inline _posts/2024-08-19-neural-sparse-v2-models.md:17]: please set featured_blog_post: false

**Review by krisfreedain — CHANGES_REQUESTED** [review]
please set featured_blog_post: false

**Review by krisfreedain — APPROVED** [review]
